extensions/amp-analytics/0.1/linker-manager.js

```diff
@@ -0,0 +1,210 @@
+    if (!linkerNames.length) {
+      return user().error(TAG,
+          'use of "linkers" requires at least one name given.');
+    }
+
+    // Each linker config has it's own set of macros to resolve.
+    this.allLinkerPromises_ = linkerNames.map(name => {
+
+      const vendorConfig = this.config_['linkers'][name];
+      const isOptIn = this.isLegacyOptIn_();
+
+      if (!isOptIn && vendorConfig['enabled'] !== true) {
+        return user().info(TAG, `linker config for ${name} is not enabled and` +
+            'will be ignored.');
+      }
+
+      const ids = vendorConfig['ids'];
+      if (!ids) {
+        return user().error(ids,
+            '"ids" is a required field for use of "linkers".');
+      }
+
+      // Keys for linker data.
+      const keys = Object.keys(ids);
+      // Expand the value of each key value pair (if necessary).
+      const valuePromises = keys.map(key => {
+        const expansionOptions = new ExpansionOptions(this.config_['vars'],
+            /* opt_iterations */ undefined,
+            /* opt_noencode */ true);
+        return this.expandTemplateWithUrlParams_(ids[key],
+            expansionOptions);
+      });
+
+      return Promise.all(valuePromises).then(values => {
+        // Rejoin each key with its expanded value.
+        const expandedIds = {};
+        values.forEach((value, i) => {
+          // Omit pair if value resolves to empty.
+          if (value) {
+            expandedIds[keys[i]] = value;
+          }
+        });
+        this.resolvedLinkers_[name] =
+            createLinker(/* version */ '1', expandedIds);
+      });
+    });
+
+    const navigation = Services.navigationForDoc(this.ampdoc_);
+    navigation.registerAnchorMutator(
+        this.linkerCallback_.bind(this), Priority.LINKER);
+  }
+
+
+  /**
+   * Expands spec using provided expansion options and applies url replacement
+   * if necessary.
+   * @param {string} template Expression that needs to be expanded.
+   * @param {!ExpansionOptions} expansionOptions Expansion options.
+   * @return {!Promise<string>} expanded template.
+   */
+  expandTemplateWithUrlParams_(template, expansionOptions) {
+    return variableServiceFor(this.ampdoc_.win)
+        .expandTemplate(template, expansionOptions)
+        .then(expanded => Services.urlReplacementsForDoc(
+            this.ampdoc_).expandUrlAsync(expanded));
+  }
+
+
+  /**
+   * If the document has existing cid meta tag they do not need to explicity
+   * opt-in to use linker.
+   * @return {boolean}
+   * @private
+   */
+  isLegacyOptIn_() {
+    const optInMeta = this.ampdoc_.win.document.head
+        ./*OK*/querySelector(`meta[name=${GOOGLE_CID_API_META_NAME}]`);
+    const isGaType = this.type_ === 'googleanalytics';
+
+    return !!(optInMeta && isGaType);
+  }
+
+  /**
+   * Called on click on any anchor element. Adds linker param if a match for
+   * given linker configuration.
+   * @param {!Element} element
+   * @private
```

Comment:
visibleForTesting

Reply:
I tried this, but if you add `@visibleForTesting` and it is used in the class it throws an error

Reply:
I think you will need to remove `@private` and also the ending `_` in method name.

Reply:
Got it. Done.